### ISSUE TITLE
Comparison between 3.8.7 and 4.x catalogs

### ISSUE THREAD

**Issue by jairojunior** [github/octocatalog-diff#96]
First of all, thanks for this awesome project. 

I've been using it since yesterday, but compiling catalogs locally the same way rspec-puppet[1] does (i.e. `Puppet::Parser::Compiler.compile`) and exporting them to JSON. I've managed to use my control-repo configuration (r10k.yaml and hiera.yaml) and ENC data from Foreman API to this.

When I have the catalogs I just need to call:

`OctocatalogDiff::CatalogDiff::Differ.new({ :ignore_tags => true, :ignore => { ... } }, catalog2, catalog1)`

It's been working just fine, except for comparisons between 3.8.7 (without future parser) and 4.x compiled catalogs, which yields a lot of:

```
File[/etc/jboss-as/jboss-as.conf] =>
 parameters =>
   notify =>
    - Service[wildfly]
    + ["Service[wildfly]"]
```

```
Apache::Vhost[default] =>
 parameters =>
   override =>
    - None
    + ["None"]
   scriptaliases =>
    - {"alias"=>"/cgi-bin", "path"=>"/var/www/cgi-bin"}
    + [{"alias"=>"/cgi-bin", "path"=>"/var/www/cgi-bin"}]
```
Couldn't find an exact pattern for this behavior, but I was able to eliminate it by introducing an ugly code at: `OctocatalogDiff::CatalogDiff::Differ#cleanse_parameters_hash`

```
result.each do |k, v|
  if v.is_a?(Array) and v.size == 1
    result[k] = v[0]
  end
end
```

Any thoughts on this?

[1] https://github.com/rodjek/rspec-puppet/blob/master/lib/rspec-puppet/adapters.rb#L69

**Comment by kpaulisse** [comment]
These are actual changes in the catalog - there is a difference between `x` and `[x]` (regardless of what data structure `x` is). I saw some of these when we did our 3.8 to 4.x upgrade as well. Some of the time these changes will be important, especially if you're using parameter type checking - which you'll need the future parser to be able to use. In any case we would not want to remove these changes from the difference list by default.

If you're looking to implement the code you've suggested in a way that we could accept into the project, it's probably best to do it as a filter.

https://github.com/github/octocatalog-diff/tree/master/lib/octocatalog-diff/catalog-diff/filter

https://github.com/github/octocatalog-diff/pull/84/files is a PR that added a new filter and command line options for it. (And also a whole lot of test fixtures, which are most of the lines added. 😸) So if you are looking to make a contribution that's the sort of thing you'll want to do.
